README.md
# UI TOOL
### Make your apps responsive.
---

With various screen sizes and pixel densities of mobile phones, it is difficult to maintain the consistency of the UI design. So, UI Tool is developed for every mobile phone screen size, which calculates the ratio of the current device compared to the originally designed device and then applies it to the Size parameters.(i.e If the UI is originally designed for iPhone 12 Pro Max, the UI on the iPhone 5s must be the similar. This may cause fat-thumb problem. Use it wisely.)


These are the Initial views of UIs without applying Calculated Ratio.
<p align='center'>
    <img src="./resources/before.png" />
</p>

After applying the Calculated Ratio
<p align='center'>
    <img src="./resources/after.png" />
</p>


## Installation

Add this to your `pubspec.yaml` dependencies:

```yml
ui_tool: ^1.0.1
```

## How to use

Add the widget to your app like this (It automatically calculates the ratio):

If the design is given with viewport sizes (iPhone 12 Pro Max):

```dart

MaterialApp(
      builder: (context, child) {
        UI.initialize(
            context,
            viewportWidth: 428,
            viewportHeight: 926,
            );
        return  MediaQuery(
            data: MediaQuery.of(context).copyWith(textScaleFactor: UI.ratio),
            child: child!,
        );
      },
)

```


Then use the ratio with your size values.

```dart

import 'package:ui_tool/ui_tool.dart';

    Container(      
        width: 200.r,
        height: 250.r,
        child:  Text(
           "Sample Responsive Text",
           style: TextStyle(
               fontSize: 16, //Fonts will be automatically scaled
           )
        ),
    ),

```


CHANGELOG.md
## 1.0.1

* Updated readme.
## 1.0.0

* Final release after testing.  
## 0.2.0

* Added ratio extension.
## 0.1.1

* Update instructions.
## 0.1.0

* Update instructions.

## 0.0.1

* UI tool to increase the UI Widget Responsiveness.


lib/src/ui.dart
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

/// The base class which keep all the necessary configs
class UI {
  /// Width
  static double width = 0;

  /// Height
  static double height = 0;

  /// Height in pixels
  static double hpx = 0;

  /// Width in pixels
  static double wpx = 0;

  /// Calculated ratio
  static double ratio = 1;

  /// Is the device in landscape orientation
  static bool isLandscape = false;

  /// is a WEB or PC or Mac
  static bool isWeb = false;

  /// device UI value
  static double deviceUIvalue = 0;

  /// Loads the text theme
  static TextTheme textTheme = const TextTheme();

  /// This will initialize the UI and calculate the ration on each screan load.
  static void initialize(context,
      {required int viewportWidth, required int viewportHeight}) {
    width = MediaQuery.of(context).size.width;
    height = MediaQuery.of(context).size.height;
    hpx = height * MediaQuery.of(context).devicePixelRatio;
    wpx = width * MediaQuery.of(context).devicePixelRatio;
    isLandscape = MediaQuery.of(context).orientation == Orientation.landscape;
    isWeb = kIsWeb;

    textTheme = Theme.of(context).textTheme;

    if (isWeb) {
      ratio = 1;
    } else {
      double designValue =
          pow((pow(viewportWidth, 2) + pow(viewportHeight, 2)), 0.5).toDouble();
      deviceUIvalue = pow((pow(width, 2) + pow(height, 2)), 0.5).toDouble();
      ratio = deviceUIvalue / designValue;
    }

    if (kDebugMode) {
      debugPrint('''
Width (viewport)  = $width
Height (viewport) = $height
Width(in pixels)  = $wpx
Height(in pixels) = $hpx
UI Value          = $deviceUIvalue
Ratio             = $ratio
        ''');
    }
  }
}

extension SizerExt on num {
  /// Calculates the sp (Scalable Pixel) depending on the device's screen size (ratio)
  double get r => this * UI.ratio;
}


lib/src/wrapper.dart
import 'package:flutter/widgets.dart';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:ui_tool/src/ui.dart';

///Wrap your first Screen with UIWrapper
/// * [child] - The first screen HomePage / SplashScreen.
/// * [width] - The width of the design
/// * [height] - The height of the design
class UIWrapper extends StatelessWidget {
  final Widget child;
  final double width;
  final double height;
  final bool isViewPort;

  const UIWrapper({
    required this.child,
    required this.width,
    required this.height,
    this.isViewPort = false,
  });
  @override
  Widget build(BuildContext context) {
    if (kIsWeb) {
      UI.isWeb = true;
    } else {
      UI.isWeb = false;
    }

    UI.initialize(
      context,
      width,
      height,
      isViewPort,
    );

    return child;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      builder: (context, child) {
        UI.initialize(
            context,
            viewportWidth: 428,
            viewportHeight: 926,
            );
        return  MediaQuery(
            data: MediaQuery.of(context).copyWith(textScaleFactor: UI.ratio),
            child: child!,
        );
      },
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;
  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Container(      
                width: 200.r,
                height: 250.r,
                child:  Text(
                  "Sample Responsive Text",
                  style: TextStyle(
                      fontSize: 16, //Fonts will be automatically scaled
                  )
                ),
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
    );
  }
}
